File: lib/server/restr-leaderboard-scoring.js
/**
 * Restaurant AI leaderboard scoring (restr_ai_leaderboard).
 *
 * Based on salon leaderboard-scoring.js with restaurant-specific Google Visibility
 * customer-acquisition emphasis:
 * - Rating band 3.5–5★ (restaurants cluster higher than salons)
 * - Review volume plateaus ~2,000 (vs ~900 for salons)
 * - Composite ai_score adds a visibility blend (local SEO + conversion dims) at 22%
 * - dim_local_seo_score: menu, hours, photos, categories, price, place_id
 * - dim_conversion_score: phone, reservations, delivery/takeout, menu URL
 *
 * Keep in sync: pipelines/restr_leaderboard_scoring.py
 */

const {
  assessmentLevelFromFractionFromTop,
  assignAssessmentLevelsByAiScorePercentile,
  freshnessHeuristic,
  resolveAssessmentLevelForSlug,
  sentimentFromReviewHistogram,
} = require("./leaderboard-scoring");

const VOLUME_PLATEAU = 2000;
const RATING_FLOOR = 3.5;

function clamp01(x, fallback) {
  const n = Number(x);
  if (!Number.isFinite(n)) return fallback;
  return Math.min(1, Math.max(0, n));
}

function round1(v) {
  return Math.round(Math.min(100, Math.max(0, Number(v) || 0)) * 10) / 10;
}

/**
 * Restaurant Google Visibility composite 0–100.
 * @param {number} rating
 * @param {number} reviewCount
 * @param {number} sentimentP
 * @param {number} freshnessF
 * @param {number} [localSeoScore] 0–100
 * @param {number} [conversionScore] 0–100
 */
function calcRestrAiScore(rating, reviewCount, sentimentP, freshnessF, localSeoScore, conversionScore) {
  const r = Math.min(5, Math.max(1, Number(rating) || 0));
  const n = Math.max(0, Math.floor(Number(reviewCount) || 0));
  const pn = clamp01(sentimentP, 0.72);
  const fn = clamp01(freshnessF, 0.7);
  const localSeo = Math.min(100, Math.max(0, Number(localSeoScore) || 0));
  const conversion = Math.min(100, Math.max(0, Number(conversionScore) || 0));

  const ratingNorm = Math.min(1, Math.max(0, (r - RATING_FLOOR) / (5 - RATING_FLOOR)));
  const volumeNorm = Math.min(1, Math.log10(n + 1) / Math.log10(VOLUME_PLATEAU));
  const visibilityNorm = (localSeo * 0.55 + conversion * 0.45) / 100;

  const blend =
    ratingNorm * 0.24 +
    volumeNorm * 0.22 +
    pn * 0.16 +
    fn * 0.1 +
    visibilityNorm * 0.28;

  const conf = 1 - Math.exp(-n / 65);
  const prior = 0.5;
  const adjusted = prior + (blend - prior) * (0.34 + 0.66 * conf);
  const out = adjusted * 100;
  return Math.round(Math.min(100, Math.max(0, out)) * 10) / 10;
}

/**
 * Six dimension scores for restaurant Google Visibility.
 * @param {{
 *   rating?: number,
 *   reviewCount?: number,
 *   sentimentP?: number,
 *   freshnessF?: number,
 *   phone?: string,
 *   website?: string,
 *   address?: string,
 *   placeId?: string,
 *   menuUrl?: string,
 *   openingHours?: unknown,
 *   imageUrl?: string,
 *   categories?: unknown,
 *   price?: string,
 *   delivery?: boolean,
 *   takeout?: boolean,
 *   dineIn?: boolean,
 *   reservable?: boolean,
 * }} signals
 */
function restrDimensionScoresFromSignals(signals) {
  const s = signals || {};
  const rating = Number(s.rating) || 0;
  const reviewCount = Math.max(0, Math.floor(Number(s.reviewCount) || 0));
  const sentimentP = clamp01(s.sentimentP, 0.8);
  const freshnessF = clamp01(s.freshnessF, 0.75);
  const has = (v) => v !== null && v !== undefined && String(v).trim().length > 0;
  const hasHours =
    (Array.isArray(s.openingHours) && s.openingHours.length > 0) ||
    (s.openingHours && typeof s.openingHours === "object" && Object.keys(s.openingHours).length > 0);
  const hasCategories = Array.isArray(s.categories) ? s.categories.length > 0 : has(s.categories);

  const ratingScore = round1(Math.max(0, ((rating - RATING_FLOOR) / (5 - RATING_FLOOR)) * 100));
  const reviewScore = round1((Math.log10(reviewCount + 1) / Math.log10(VOLUME_PLATEAU)) * 100);
  const sentimentScore = round1(sentimentP * 100);
  const recencyScore = round1(freshnessF * 100);

  const localSeoScore = round1(
    (has(s.website) ? 14 : 0) +
      (has(s.menuUrl) ? 22 : 0) +
      (hasHours ? 16 : 0) +
      (has(s.imageUrl) ? 10 : 0) +
      (hasCategories ? 8 : 0) +
      (has(s.price) ? 10 : 0) +
      (has(s.address) ? 8 : 0) +
      (has(s.placeId) ? 12 : 0),
  );

  const conversionScore = round1(
    ratingScore * 0.18 +
      reviewScore * 0.14 +
      (has(s.phone) ? 22 : 0) +
      (has(s.menuUrl) ? 18 : 0) +
      (s.reservable ? 12 : 0) +
      (s.delivery || s.takeout ? 10 : 0) +
      (s.dineIn ? 6 : 0) +
      (has(s.placeId) ? 10 : 0),
  );

  return {
    dim_reviews_score: reviewScore,
    dim_rating_score: ratingScore,
    dim_sentiment_score: sentimentScore,
    dim_recency_score: recencyScore,
    dim_local_seo_score: localSeoScore,
    dim_conversion_score: conversionScore,
  };
}

module.exports = {
  RATING_FLOOR,
  VOLUME_PLATEAU,
  assessmentLevelFromFractionFromTop,
  assignAssessmentLevelsByAiScorePercentile,
  calcRestrAiScore,
  freshnessHeuristic,
  resolveAssessmentLevelForSlug,
  restrDimensionScoresFromSignals,
  sentimentFromReviewHistogram,
};

File: pipelines/restr_leaderboard_scoring.py
"""
Mirror lib/server/restr-leaderboard-scoring.js for restr_ai_leaderboard ingestion.
"""

from __future__ import annotations

import math
from typing import Any, Dict, Mapping, Optional

from leaderboard_scoring import (
    assign_assessment_levels_by_ai_score_percentile,
    freshness_heuristic,
    sentiment_from_review_histogram,
)

VOLUME_PLATEAU = 2000
RATING_FLOOR = 3.5


def _clamp01(x: Any, fallback: float = 0.8) -> float:
    try:
        n = float(x)
    except (TypeError, ValueError):
        return fallback
    return max(0.0, min(1.0, n))


def _round1(v: float) -> float:
    return round(max(0.0, min(100.0, float(v))), 1)


def calc_restr_ai_score(
    rating: float,
    review_count: int,
    sentiment_p: float,
    freshness_f: float,
    local_seo_score: float = 0.0,
    conversion_score: float = 0.0,
) -> float:
    r = max(1.0, min(5.0, float(rating or 0)))
    n = max(0, int(review_count or 0))
    pn = _clamp01(sentiment_p, 0.72)
    fn = _clamp01(freshness_f, 0.7)
    local_seo = max(0.0, min(100.0, float(local_seo_score or 0)))
    conversion = max(0.0, min(100.0, float(conversion_score or 0)))

    rating_norm = max(0.0, min(1.0, (r - RATING_FLOOR) / (5.0 - RATING_FLOOR)))
    volume_norm = min(1.0, math.log10(n + 1) / math.log10(VOLUME_PLATEAU))
    visibility_norm = (local_seo * 0.55 + conversion * 0.45) / 100.0

    blend = (
        rating_norm * 0.24
        + volume_norm * 0.22
        + pn * 0.16
        + fn * 0.10
        + visibility_norm * 0.28
    )
    conf = 1.0 - math.exp(-n / 65.0)
    prior = 0.5
    adjusted = prior + (blend - prior) * (0.34 + 0.66 * conf)
    return _round1(adjusted * 100.0)


def restr_dimension_scores_from_signals(signals: Mapping[str, Any]) -> Dict[str, float]:
    rating = float(signals.get("rating") or 0)
    review_count = max(0, int(signals.get("reviewCount") or signals.get("review_count") or 0))
    sentiment_p = _clamp01(signals.get("sentimentP") or signals.get("sentiment_p"), 0.8)
    freshness_f = _clamp01(signals.get("freshnessF") or signals.get("freshness_f"), 0.75)

    def has_val(v: Any) -> bool:
        return v is not None and str(v).strip() != ""

    opening = signals.get("openingHours") or signals.get("opening_hours")
    has_hours = bool(opening) and (
        (isinstance(opening, list) and len(opening) > 0)
        or (isinstance(opening, dict) and len(opening) > 0)
    )
    categories = signals.get("categories")
    has_categories = (isinstance(categories, list) and len(categories) > 0) or has_val(categories)

    rating_score = _round1(max(0.0, ((rating - RATING_FLOOR) / (5.0 - RATING_FLOOR)) * 100.0))
    review_score = _round1((math.log10(review_count + 1) / math.log10(VOLUME_PLATEAU)) * 100.0)
    sentiment_score = _round1(sentiment_p * 100.0)
    recency_score = _round1(freshness_f * 100.0)

    local_seo_score = _round1(
        (22 if has_val(signals.get("menuUrl") or signals.get("menu_url")) else 0)
        + (14 if has_val(signals.get("website")) else 0)
        + (16 if has_hours else 0)
        + (10 if has_val(signals.get("imageUrl") or signals.get("image_url")) else 0)
        + (8 if has_categories else 0)
        + (10 if has_val(signals.get("price")) else 0)
        + (8 if has_val(signals.get("address")) else 0)
        + (12 if has_val(signals.get("placeId") or signals.get("place_id")) else 0)
    )

    conversion_score = _round1(
        rating_score * 0.18
        + review_score * 0.14
        + (22 if has_val(signals.get("phone")) else 0)
        + (18 if has_val(signals.get("menuUrl") or signals.get("menu_url")) else 0)
        + (12 if signals.get("reservable") else 0)
        + (10 if (signals.get("delivery") or signals.get("takeout")) else 0)
        + (6 if signals.get("dineIn") or signals.get("dine_in") else 0)
        + (10 if has_val(signals.get("placeId") or signals.get("place_id")) else 0)
    )

    return {
        "dim_reviews_score": review_score,
        "dim_rating_score": rating_score,
        "dim_sentiment_score": sentiment_score,
        "dim_recency_score": recency_score,
        "dim_local_seo_score": local_seo_score,
        "dim_conversion_score": conversion_score,
    }
